immutablecollections/_immutableset.py

```diff
@@ -231,10 +232,20 @@ def difference(self, other: AbstractSet[Any]) -> "ImmutableSet[T]":
 
     def __sub__(self, other: AbstractSet[Any]) -> "ImmutableSet[T]":
         """
-        Gets a new set with all items in this set not in the other.
+        Subtracts other from self. An item is included iff it is in self and not in other.
         """
         return self.difference(other)
 
+    def __rsub__(self, other: AbstractSet[V]) -> "AbstractSet[V]":
+        """
+        Subtracts self from other. An item is included it is in other and not in self.
+
+        This method is necessary because it defines behavior for an `otherset - immutableset`.
+        Without this method, otherset.__sub__(immutableset) would be called, which is typically
```

Comment:
Is this accurate? I think `__sub__` typically is defined; the problem is that it tried to construct the result by passing the items directly to the `immutableset` constructor, which fails.

Reply:
That is correct. Here is the default method:
```python
    def __rsub__(self, other):
        if not isinstance(other, Set):
            if not isinstance(other, Iterable):
                return NotImplemented
            other = self._from_iterable(other)
        return self._from_iterable(value for value in other
                                   if value not in self)
```
```
In [31]: {1, 2, 3} - immutableset([3, 4, 5])
---------------------------------------------------------------------------
TypeError                                 Traceback (most recent call last)
<ipython-input-31-e63d2b5f9554> in <module>
----> 1 {1, 2, 3} - immutableset([3, 4, 5])

~/.conda/envs/aida_tools/lib/python3.6/_collections_abc.py in __rsub__(self, other)
    507                 return NotImplemented
    508             other = self._from_iterable(other)
--> 509         return self._from_iterable(value for value in other
    510                                    if value not in self)
    511

~/.conda/envs/aida_tools/lib/python3.6/_collections_abc.py in _from_iterable(cls, it)
    470         does not accept an iterable for an input.
    471         '''
--> 472         return cls(it)
    473
    474     def __and__(self, other):

TypeError: __init__() missing 2 required positional arguments: 'iteration_order' and 'top_level_type'
```
```
In [42]: type(ims)
Out[42]: immutablecollections._immutableset.FrozenSetBackedImmutableSet

In [43]: type(ims)(['a', 'c', 'b'])
---------------------------------------------------------------------------
TypeError                                 Traceback (most recent call last)
<ipython-input-43-8c89ba4f466f> in <module>
----> 1 type(ims)(['a', 'c', 'b'])

TypeError: __init__() missing 2 required positional arguments: 'iteration_order' and 'top_level_type'
```

Reply:
Strange, why is `set.__rsub__` being called when the normal set is on the LHS?